### PR TITLE
Add VerifyChannelUpgradeSequence method + tests

### PR DESCRIPTION
**Pull request by chatton**
## Description

closes: #1608

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

### CODE REVIEW

File: modules/core/03-connection/keeper/verify.go [hunk 1]
```diff
@@ -364,6 +364,44 @@ func (k Keeper) VerifyNextSequenceRecv(
...
+	if err != nil {
+		return err
+	}
+
+	if err := clientState.VerifyMembership(
+		ctx, clientStore, k.cdc, height,
+		0, 0, // skip delay period checks for non-packet processing verification
+		proof, merklePath, sdk.Uint64ToBigEndian(upgradeSequence),
+	); err != nil {
+		return errorsmod.Wrapf(err, "failed upgrade sequence receive verification for client (%s)", clientID)
```
Comment (damiannolan) [inline modules/core/03-connection/keeper/verify.go:399]:
```suggestion
		return errorsmod.Wrapf(err, "failed upgrade sequence verification for client (%s)", clientID)
```

**Review by damiannolan — APPROVED** [review]
Looks good to me!

File: modules/core/03-connection/keeper/verify.go [hunk 2]
```diff
@@ -364,6 +364,44 @@ func (k Keeper) VerifyNextSequenceRecv(
...
 }
 
+// VerifyChannelUpgradeSequence verifies a proof of the upgrade sequence number to be
+// used during channel upgrades.
+func (k Keeper) VerifyChannelUpgradeSequence(
+	ctx sdk.Context,
+	connection exported.ConnectionI,
+	height exported.Height,
+	proof []byte,
+	portID,
```
Comment (charleenfei) [inline modules/core/03-connection/keeper/verify.go:374]: should these be `counterpartyPortID` and `counterpartyChannelID` as per the spec? just to be explicit

**Review by charleenfei — APPROVED** [review]
lgtm!

File: modules/core/03-connection/keeper/verify_test.go [hunk 3]
```diff
@@ -682,6 +682,89 @@ func (suite *KeeperTestSuite) TestVerifyNextSequenceRecv() {
...
+
+			err := path.EndpointA.ChanUpgradeInit(path.EndpointB.Chain.GetTimeoutHeight(), 0)
+			suite.Require().NoError(err)
+
+			tc.malleate()
+
+			upgradeSequenceKey := host.ChannelUpgradeSequenceKey(path.EndpointA.ChannelConfig.PortID, path.EndpointA.ChannelID)
+			proof, proofHeight := suite.chainA.QueryProof(upgradeSequenceKey)
+
+			err = suite.chainB.GetSimApp().IBCKeeper.ConnectionKeeper.VerifyChannelUpgradeSequence(suite.chainB.GetContext(), path.EndpointB.GetConnection(), malleateHeight(proofHeight, 0), proof, path.EndpointB.ChannelConfig.PortID, path.EndpointB.ChannelID, expectedUpgradeSequence)
```
Comment (crodriguezvega) [inline modules/core/03-connection/keeper/verify_test.go:757]: Same nit as [here](https://github.com/cosmos/ibc-go/pull/3342#discussion_r1149705072) :)

**Review by crodriguezvega — APPROVED** [review]
LGTM!

File: modules/core/03-connection/keeper/verify_test.go (hunk 3 above)
Comment (chatton) [reply modules/core/03-connection/keeper/verify_test.go]: it is almost certainly not required!

**Comment by damiannolan** [comment]
LFG

**Pull request merged**